src/odemis/acq/stream_settings.py

```diff
@@ -0,0 +1,174 @@
+
+This file is part of Odemis.
+
+Odemis is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License version 2 as published by the Free Software Foundation.
+
+Odemis is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with Odemis. If not, see http://www.gnu.org/licenses/.
+"""
+import json
+import logging
+import os
+from typing import List, Dict, Any
+
+from odemis import model, util
+from odemis.acq.stream import FluoStream
+from odemis.util.img import md_format_to_tint, tint_to_md_format
+
+SETTINGS_ORDER = ["name", "excitation", "power", "emission", "tint", "detExposureTime", "auto_bc_outliers"]
+NON_SETTINGS_VA = ['acquisitionType', 'auto_bc', 'background', 'histogram', 'image', 'intensityRange', 'is_active',
+                   'roi', 'should_update', 'single_frame_acquisition', 'status']
+
+
+def get_config_index(config_data: list, name: str) -> int:
+    """
+    Get the index of the config entry with the given name
+    :param config_data: list of config entries
+    :param name: name of config entry
+    :return: index of config entry with given name
+    """
+    try:
+        index = list(map(lambda x: x["name"], config_data)).index(name)
+    except:
+        index = None
+
+    return index
+
+
+class StreamSettings:
+    # TODO change the description
+    """
+    Read and write a list of most recently used (mru) settings from and to a JSON file respectively. The first element
+    in the list will be the latest used setting and the last element will be the oldest used setting. The list will
+    consist of finite number of entries which will update in the above mention order each time the list of settings
+    is updated.
+    """
+
+    def __init__(self, file_path, max_entries, config_data: List[Dict[str, Any]] = None):
+        """
+        :param file_path: full directory path to save the JSON file
+        :param max_entries: number of entries to save in the JSON file
+        :param config_data: list of stream controller settings where the first element is the latest
+         stream entry. The number of entries is defined by max_entries.
+        """
+        self.max_entries = max_entries
+        self.config_data = config_data
+        self.file_path = file_path
+        self._read()  # update the config_data by reading the JSON file
+        self.settings_order = None  # the ordered list of setting parameters used for loading any given stream
+
+    def _read(self):
+        logging.debug("Reading the most recently used stream settings in JSON")
+        if os.path.exists(self.file_path):
+            try:
+                with open(self.file_path, 'r') as jsonfile:
+                    self.config_data = json.load(jsonfile)
```

Comment:
There is a lot of JSON files that would not be of the right structure. Please check that config_data is a List[Dict[str,Any]], and if not, log a warning and reset config_data to `[]`.